storage/grpc_client.go

```diff
@@ -2370,3 +2382,78 @@ func checkCanceled(err error) error {
 
 	return err
 }
+
+func (w *gRPCWriter) initializeRetryConfig() {
+	if w.attempts == 0 {
+		w.attempts = 1
+
+		if w.settings.retry == nil {
+			w.settings.retry = defaultRetry
+		}
+
+		if w.settings.retry.shouldRetry == nil {
+			w.settings.retry.shouldRetry = ShouldRetry
+		}
+
+		w.invocationID = uuid.New().String()
+
+		if w.settings.retry.backoff != nil {
+			w.backoff.Multiplier = w.settings.retry.backoff.Multiplier
+			w.backoff.Initial = w.settings.retry.backoff.Initial
+			w.backoff.Max = w.settings.retry.backoff.Max
+		}
+	}
+}
+
+// shouldRetry determines if a retry is necessary and if so waits the appropriate
+// amount of time. It returns true if the error is retryable or the error to be
+// surfaced to the user if not.
+func (w *gRPCWriter) shouldRetry(ctx context.Context, err error) (bool, error) {
+	if err == nil {
+		// a nil err does not need to be retried
+		return false, nil
+	}
+	if err != context.Canceled && err != context.DeadlineExceeded {
+		w.lastErr = err
+	}
+
+	retryConfig := w.settings.retry
+
+	if retryConfig.policy == RetryNever {
+		return false, err
+	}
+
+	if retryConfig.maxAttempts != nil && w.attempts >= *retryConfig.maxAttempts {
+		return false, fmt.Errorf("storage: retry failed after %v attempts; last error: %w", w.attempts, err)
+	}
+
+	w.attempts++
+
+	retryable := retryConfig.shouldRetry(err)
+	// Explicitly check context cancellation so that we can distinguish between a
+	// DEADLINE_EXCEEDED error from the server and a user-set context deadline.
+	// Unfortunately gRPC will codes.DeadlineExceeded (which may be retryable if it's
+	// sent by the server) in both cases.
+	ctxErr := ctx.Err()
+	if errors.Is(ctxErr, context.Canceled) || errors.Is(ctxErr, context.DeadlineExceeded) {
+		retryable = false
+
+		if w.lastErr != nil {
+			return false, fmt.Errorf("retry failed with %v; last error: %w", ctxErr, w.lastErr)
+		}
+		return false, ctxErr
+	}
+
+	if retryable {
+		p := w.backoff.Pause()
```

Comment:
As discussed, maybe we can add a mock test in this PR to ensure that backoff.Pause is actually called?

Reply:
I made a mock for the backoff and had to change more than anticipated to be able to inject it, PTAL